dids/src/main/kotlin/web5/sdk/dids/methods/dht/DidDht.kt

```diff
@@ -45,393 +79,500 @@ public class DidDhtConfiguration internal constructor(
+                .controller(URI.create(did))
+                .publicKeyJwk(publicKeyJwk.toPublicJWK().toJSONObject())
+                .build()
+
+              keyLookup[name.split(".")[0].drop(1)] = "$did#$vmID"
+            }
+            // handle services
+            name.startsWith("_s") -> {
+              val data = parseTxtData(rr.strings.joinToString(","))
+              services += Service.builder()
+                .id(URI.create("$did#${data["id"]!!}"))
+                .type(data["t"]!!)
+                .serviceEndpoint(data["uri"]!!)
+                .build()
+            }
+            // handle type indexing
+            name == "_typ._did." -> {
+              if (rr.strings[0].isNotEmpty() && rr.strings.size == 1) {
+                types += rr.strings[0].removePrefix("id=").split(",").map {
+                  DidDhtTypeIndexing.fromInt(it.toInt()) ?: throw IllegalArgumentException("invalid type index")
                 }
+              } else {
+                throw IllegalArgumentException("invalid types record")
               }
-              // handle root record
-              name == "_did." -> {
-                val rootData = rr.strings.joinToString(";").split(";")
-
-                val lists = mapOf(
-                  "auth" to mutableListOf<VerificationMethod>(),
-                  "asm" to mutableListOf(),
-                  "agm" to mutableListOf(),
-                  "inv" to mutableListOf(),
-                  "del" to mutableListOf()
-                )
-
-                rootData.forEach { item ->
-                  val (key, values) = item.split("=")
-                  val valueItems = values.split(",")
-
-                  valueItems.forEach {
-                    lists[key]?.add(VerificationMethod.builder().id(URI(keyLookup[it]!!)).build())
-                  }
-                }
+            }
+            // handle root record
+            name == "_did." -> {
+              val rootData = rr.strings.joinToString(";").split(";")
+
+              val lists = mapOf(
+                "auth" to mutableListOf<VerificationMethod>(),
+                "asm" to mutableListOf(),
+                "agm" to mutableListOf(),
+                "inv" to mutableListOf(),
+                "del" to mutableListOf()
+              )
+
+              rootData.forEach { item ->
+                val (key, values) = item.split("=")
+                val valueItems = values.split(",")
 
-                // add verification relationships
-                doc.authenticationVerificationMethods(lists["auth"])
-                doc.assertionMethodVerificationMethods(lists["asm"])
-                doc.keyAgreementVerificationMethods(lists["agm"])
-                doc.capabilityInvocationVerificationMethods(lists["inv"])
-                doc.capabilityDelegationVerificationMethods(lists["del"])
+                valueItems.forEach {
+                  lists[key]?.add(VerificationMethod.builder().id(URI(keyLookup[it]!!)).build())
```

Comment:
Can keyLookup be null?

Reply:
yes that's why there is `!!`

Reply:
this method does a best effort to reconstruct a DID document, it'll fail if it's not a properly constructed packet